### PR TITLE
Fixes #73 - (MB-33014) Support arithmetic operations

### LINKED ISSUE #73: Matcher to support simple arithmetic
Since we currently support various math functions, it would make sense to add support to simple math arithmetic operations, as these are supported by N1QL.
https://docs.couchbase.com/server/6.0/n1ql/n1ql-language-reference/arithmetic.html

These (+ - / * % -) would translate nicely into functions part of the match tree.

### PR DESCRIPTION
**Pull request by nelio2k**

### CODE REVIEW

File: constants.go [hunk 1]
```diff
@@ -30,6 +30,12 @@ const (
 	MathFuncSin     string = "mathSin"
 	MathFuncSqrt    string = "mathSqrt"
 	MathFuncTan     string = "mathTan"
+	MathFuncAdd     string = "mathAdd"
+	MathFuncSub     string = "mathSubract"
+	MathFuncMult    string = "mathMultiply"
```
Comment (brett19) [inline constants.go]: Why not `MathFuncMul`?

File: expression.go [hunk 2]
```diff
@@ -255,3 +255,47 @@ type LikeExpr struct {
...
+type MathAddExpr struct {
+	Lhs Expression
+	Rhs Expression
+}
+
+func (expr MathAddExpr) String() string {
+	return fmt.Sprintf("%s + %s", expr.Lhs, expr.Rhs)
+}
+
+type MathDelExpr struct {
```
Comment (brett19) [inline expression.go]: `MathSubExpr`?

File: fastMatcher.go [hunk 3]
```diff
@@ -166,6 +166,29 @@ func (m *FastMatcher) resolveFunc(fn FuncRef, activeLit *FastVal) FastVal {
...
+		p2 := m.resolveParam(fn.Params[1], activeLit)
+		return FastValMathAdd(p1, p2)
+	case MathFuncSub:
+		p1 := m.resolveParam(fn.Params[0], activeLit)
+		p2 := m.resolveParam(fn.Params[1], activeLit)
+		return FastValMathSub(p1, p2)
+	case MathFuncMult:
+		p1 := m.resolveParam(fn.Params[0], activeLit)
+		p2 := m.resolveParam(fn.Params[1], activeLit)
+		return FastValMathMult(p1, p2)
```
Comment (brett19) [inline fastMatcher.go]: `FastValMathMul`?

File: fastMatcher.go [hunk 4]
```diff
@@ -166,6 +166,29 @@ func (m *FastMatcher) resolveFunc(fn FuncRef, activeLit *FastVal) FastVal {
...
+		p1 := m.resolveParam(fn.Params[0], activeLit)
+		p2 := m.resolveParam(fn.Params[1], activeLit)
+		return FastValMathDiv(p1, p2)
+	case MathFuncMod:
+		p1 := m.resolveParam(fn.Params[0], activeLit)
+		p2 := m.resolveParam(fn.Params[1], activeLit)
+		return FastValMathMod(p1, p2)
+	case MathFuncNeg:
+		p1 := m.resolveParam(fn.Params[0], activeLit)
+		return FastValMathNegate(p1)
```
Comment (brett19) [inline fastMatcher.go]: Why not use `Neg` rather than `Negate` here to stay consistent?

File: filterExprParser.go [hunk 5]
```diff
@@ -622,6 +679,71 @@ func (f *FEOnePathFuncNoArgName) OutputExpression() (Expression, error) {
 	return nil, fmt.Errorf("Not supported (FEOnePathFuncNoArgName) %v", f.String())
 }
 
+type FEMathArithmeticOp struct {
+	Addition *bool `@"+" |`
+	Deletion *bool `@"-" |`
```
Comment (ysui6888) [inline filterExprParser.go]: subtraction?

File: expression.go (hunk 2 above)
Comment (nelio2k) [reply expression.go]: Actually these aren't really needed

File: constants.go (hunk 1 above)
Comment (nelio2k) [reply constants.go]: Sure

File: fastMatcher.go (hunk 3 above)
Comment (nelio2k) [reply fastMatcher.go]: Sure

File: fastMatcher.go (hunk 4 above)
Comment (nelio2k) [reply fastMatcher.go]: Sure

File: filterExprParser.go (hunk 5 above)
Comment (nelio2k) [reply filterExprParser.go]: Sure

**Comment by nelio2k** [comment]
> is the fix trying to add new functionality or to make the rules more specific?

Both... the rules need to be made more specific for capturing, and then the internal fastMatcher needs to add the operations to do the actual arithmetic.

**Pull request merged**